lib/experimental/class-gutenberg-rest-template-revision-count.php

```diff
@@ -0,0 +1,72 @@
+<?php
+/**
+ * WP_REST_Templates_Controller class
+ *
+ * When merging into core, prepare_item_for_response() and get_item_schema()
+ * should be merged with the parent methods
+ *
+ * @package gutenberg
+ */
+
+ class Gutenberg_REST_Template_Revision_Count extends WP_REST_Templates_Controller {
+	/**
+	 * Add revisions to the response.
+	 *
+	 * @param WP_Block_Template $item    Template instance.
+	 * @param WP_REST_Request   $request Request object.
+	 * @return WP_REST_Response Response object.
+	 */
+	public function prepare_item_for_response( $item, $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		$template = $item;
+
+		$response = parent::prepare_item_for_response( $item, $request );
+		$data = $response->get_data();
+		$fields = $this->get_fields_for_response( $request );
+
+		if ( ! is_null( $template->wp_id ) && $template->wp_id !== 0 ) {
+			$revisions = wp_get_latest_revision_id_and_total_count( $template->wp_id );
+		} else {
+			// TODO: Should this be included if it's a file-based template?
+			$revisions = array(
+				'count' => 0,
+				'latest_id' => 0,
+			);
+		}
+
+		if ( rest_is_field_included( 'revision_count', $fields ) ) {
+			$data['revision_count'] = (int) $revisions['count'];
+		}
+
+		if ( rest_is_field_included( 'latest_revision_id', $fields ) ) {
+			$data['latest_revision_id'] = (int) $revisions['latest_id'];
+		}
```

Comment:
Instead of adding the data like this, please add like other endpoints, like [this](https://github.com/WordPress/wordpress-develop/blob/2517d5452561b1171be01ea7d05c1cc1cd68fe9b/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2065-L2081).

Reply:
The other endpoints provide links to REST routes that provide revision information, which I inferred as the reason that they are in the `_links` section in the response. I've noted in the How section of my description and [this comment](https://github.com/WordPress/gutenberg/pull/45215#issuecomment-1290942806) that the REST routes for revisions are not compatible with the routes for templates, because template IDs are non-numeric and use slashes internally.

I chose not to put these in the `_links` section because of the difficulty getting the Revision REST controller routes to respond. Are you saying you'd still like to have these data in `_links` without URLs for other REST responses?

Reply:
No, I would like to the revisions endpoints working.

Reply:
There is no such idea of revisioned templates for file based templates. There will only ever work with ones stored in the database. As we check to see if there are WordPress id based templates. So this is really an issue.

Reply:
What do you think about having a different URL structure for templates? Something like `/wp/v2/revisions/template[_part]s/123/456`? It seems like the revision controller could have a special case for the route for those post types to avoid conflicts.

Reply:
There is already an existing endpoint that looks like this. 

```
/wp/v2/templates/(?P<parent>[\\d]+)/revisions":
```

and 

```
/wp/v2/templates/(?P<parent>[\\d]+)/revisions/(?P<id>[\\d]+)":
```

There is no need for a  different endpoint / structure.

Reply:
Those endpoints do exist, however the routes in `WP_REST_Templates_Controller` also match those URLs and so `WP_REST_Revisions_Controller` never sees requests for them. You can confirm this by making a request to those endpoints and seeing that the response is from the Templates controller: the error code is `rest_template_not_found`.

Here is the best solution I've been able to come up with to route `/wp/v2/template-parts/18/revisions` and `/wp/v2/template-parts/123/revisions/456` to `WP_REST_Revisions_Controller`:

First, change [the route for `WP_REST_Templates_Controller::get_item`](https://github.com/WordPress/WordPress/blob/72ae8d87ac7b8feb864bda854b7a645c915d530d/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L101-L109) to add `(?<!revisions)` at the end. This prevents the controller from matching requests for `/wp/v2/template-parts/123/revisions`. This fixes requests for `/wp/v2/template-parts/18/revisions`.

This regex will still attempt to match `/wp/v2/template-parts/123/revisions/456` because that is indistinguishable from `/wp/v2/template-parts/themedir/twentytwentythree/footer` (we need to [allow for theme dirs](https://github.com/WordPress/WordPress/blob/72ae8d87ac7b8feb864bda854b7a645c915d530d/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L104), and we have to accept [a single slash between theme name and part](https://github.com/WordPress/WordPress/blob/72ae8d87ac7b8feb864bda854b7a645c915d530d/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L189-L192)). `(?<!revisions)` is a negative lookahead, and they must be fixed width so we can't add `\d+` in there to match one-or-more digits.

We may be able to update the regex for the template routes to require at least one non-numeric character in theme and template names **if** we can guarantee no one is going to name a theme `2023` or a template part `1`. If not, I put together this pseudocode that feels incredibly hacky, and not a solution I like, but it should allow us to have `WP_REST_Templates_Controller` intercept requests for `/wp/v2/template-parts/123/revisions/456`:

```php
function wp_api_reroute_template_revision_requests( $dispatch_result, $request, $route, $handler ) {
	// Check if the request is for a template revision.
	if ( is_a( $handler['callback'][0], 'WP_REST_Templates_Controller', true ) ) {
		$post_type        = $handler['callback'][0]->getPostType();
		$post_type_object = get_post_type_object( $post_type );
		$post_base        = ! empty( $post_type_object->rest_base ) ? $post_type_object->rest_base : $post_type_object->name;

		// See if we have a request for get_item()
		if ( preg_match( '@/' . $post_base . '/(?P<parent>[\d]+)/revisions/(?P<id>[\d]+)$@',
		                 $request->get_route(),
						 $get_items_matches ) ) {
			$revision_controller = new WP_REST_Revisions_Controller( $post_type );
			return $revision_controller->get_item( $get_items_matches );
		}
	}
}
add_filter( 'rest_dispatch_request', 'wp_api_reroute_template_revision_requests', 10, 4 );
```

This feels very brittle and indirect, and not a particularly elegant solution, but it is a solution nonetheless.

@spacedmonkey I would appreciate your thoughts on the proposed solutions (don't link to revisions, change the URL structure for template revisions, intercept requests for revisions in a filter), and if you have any other ideas on ensuring that the Revision controller gets the requests instead of the Template controller. Thanks again for your work on this.